### PR TITLE
Letify: use a lambda visitor

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- n/a My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by kroening** [comment]
Look at last commit only!

File: src/solvers/smt2/letify.cpp [hunk 1]
```diff
@@ -83,12 +83,17 @@ exprt letifyt::substitute_let(const exprt &expr, const seen_expressionst &map)
   if(expr.operands().empty())
     return expr;
 
-  let_visitort lv(map);
-
   exprt tmp = expr;
 
   for(auto &op : tmp.operands())
-    op.visit(lv);
+    op.visit([&map](exprt &expr) {
```
Comment (romainbrenguier) [inline src/solvers/smt2/letify.cpp:90]: multi-line `for` body should be inside braces

File: src/solvers/smt2/letify.cpp [hunk 2]
```diff
@@ -83,12 +83,17 @@ exprt letifyt::substitute_let(const exprt &expr, const seen_expressionst &map)
...
+      seen_expressionst::const_iterator it = map.find(expr);
+      if(it != map.end() && it->second.count >= letifyt::LET_COUNT)
+      {
+        const symbol_exprt &symb = it->second.let_symbol;
+        expr = symb;
```
Comment (romainbrenguier) [inline src/solvers/smt2/letify.cpp]: the temporary `symb` variable doesn't seem necessary

**Review by romainbrenguier — APPROVED** [review]
I only looked at the last commit

**Review by smowton — APPROVED** [review]
Looks fine except as @romainbrenguier notes

**Pull request merged**